### PR TITLE
fix(artwork): re-resolve artwork when image files change on disk

### LINKED ISSUE #5469: [Bug]: Quick scan does not detect new cover art
### I confirm that:

- [x] I have searched the existing [open AND closed issues](https://github.com/navidrome/navidrome/issues?q=is%3Aissue) to see if an issue already exists for the bug I've encountered
- [x] I'm using the latest version (your issue may have been fixed already)

### Version

0.61.2

### Current Behavior

After placing new or updated cover art (cover.jpg, etc) and then performing a quick scan, there is still no cover art shown.

### Expected Behavior

I would expect at least new files to be detected with a quick scan, but preferably also files with changed size or mtime.

### Steps To Reproduce

1. Add a new cover.jpg image to an existing album without cover art, or update an existing cover art image.

### Environment

```markdown
- OS: Debian Trixie
- Browser: Chromium
- Client: Web, others
```

### How Navidrome is installed?

Package

### Configuration

```toml
DataFolder = "/var/lib/navidrome"
MusicFolder = "/opt/navidrome/music"
```

### Relevant log output

```shell

```

### Anything else?

Doing a full scan does seem to detect the cover art change, but that is a way too heavy on large libraries when I only want to update one (or a few) album cover art.

### Code of Conduct

- [x] I agree to follow Navidrome's Code of Conduct

### PR DESCRIPTION
**Pull request by deluan**
### Description

When an image file was changed on disk without touching any audio file (replacing a `cover.jpg`, adding or deleting an `artist.jpg`), the scanner detected the folder change but never enqueued artwork re-resolution: the enqueue only happened for albums/artists derived from re-imported tracks. Clients polling the API kept seeing the old `coverArt` hash until something else forced a re-resolve (e.g. opening the album in the web UI, which trips the serve-time mtime check). Reported in discussion #5960.

This PR makes phase 1 diff each changed folder's image list and `imagesUpdatedAt` against the previously persisted folder state (carried through the existing `GetFolderUpdateInfo` bulk pre-pass, so no extra per-folder queries), and at the end of the phase bulk-enqueue re-resolution for the affected entities:

- **Albums**: albums with tracks in the changed folder or its direct children (covers sitting in an album folder above disc subfolders).
- **Artists**: when an artist-pattern image is involved, artists with albums anywhere under the changed folder's subtree — mirroring the resolver's upward search from the artist folder to the library root. The artist-image classification is exported from `core/artwork` (`IsArtistImageFile`) so the scanner shares the resolver's `ArtistArtPriority` token grammar instead of re-parsing it, and the album selection shares the resolver's sole-album-artist rule (`model.SoleAlbumArtistFilter`).

Because the diff also fires on deletions, this closes most of the "losing a higher-priority local image is never retried" gap: deleting a `cover.jpg` now falls back to the next source (e.g. embedded art) on the next scan.

New repository helpers keep the mapping set-based and light: `Folder.GetSubtreeIDs` (path-prefix subtree with LIKE escaping and expression-tree batching), `MediaFile.GetAlbumIDsByFolder` (distinct, indexed by `folder_id`), and `Album.GetSoleAlbumArtistIDs`. No schema change and no new index.

### Related Issues

Related to #5960

### Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Documentation update
- [ ] Refactor
- [ ] Other (please describe):

### Checklist

- [x] My code follows the project's coding style
- [x] I have tested the changes locally
- [ ] I have added or updated documentation as needed
- [x] I have added tests that prove my fix/feature works (or explain why not)
- [x] All existing and new tests pass

### How to Test

1. Point a scratch instance at a small library with folder covers and an `artist.jpg`, let the initial scan and artwork queue settle.
2. Replace the bytes of a `cover.jpg` (keep the name), trigger a quick scan via `startScan`, and fetch the album via `getAlbumList2`/`getAlbum`: the `coverArt` id's hash suffix must change without ever opening the web UI. Same for `artist.jpg` and `getArtists`.
3. Delete a cover and rescan: the album must fall back to its next source (embedded art or placeholder) with a new hash.
4. Add a shared artist image at the library root and rescan: artists without their own image pick it up; artists with their own image keep theirs.
5. Rescan (quick and full) with no changes: no hashes change and nothing is re-enqueued.

All of the above was verified against a live scratch server (12-case matrix, including multi-disc layouts and an unrelated-file over-trigger check), plus new scanner integration tests and repository tests.

### Additional Notes

- A folder change that only adds a non-image file next to an existing image still counts as an image change (`imagesUpdatedAt` folds in the folder's modTime, a known pre-existing over-trigger). The cost is an internal re-resolve that produces the same content hash, so nothing is client-visible.
- During live verification I found a pre-existing, unrelated gap: a brand-new artist imported via a *quick* scan can settle with absent artwork because the scanner skips new folders with no files, so the artist folder row needed by the resolver's climb doesn't exist yet. A full scan heals it. Will file separately.

### CODE REVIEW

**Comment by deluan** [comment]
@codex review

File: persistence/folder_repository.go [hunk 1]
```diff
@@ -172,23 +179,67 @@ func (r folderRepository) getFolderUpdateInfoBatch(lib model.Library, targetPath
...
 	return m, nil
 }
 
+// GetSubtreeIDs returns the IDs of the folders at the given library-relative paths and all their
+// descendants. A path of "" or "." selects the whole library.
+func (r folderRepository) GetSubtreeIDs(lib model.Library, paths ...string) ([]string, error) {
+	where := And{Eq{"library_id": lib.ID}, Eq{"missing": false}}
+	conds := make(Or, 0, len(paths)*3)
+	for _, p := range paths {
+		cleanPath := filepath.Clean(strings.TrimPrefix(p, string(os.PathSeparator)))
```
[the remark this answers is not in the document]
Comment (deluan) [reply persistence/folder_repository.go]: Confirmed and fixed in 7e2b1fe66. The scanner walks an io/fs, so these paths are always slash-form, but `filepath.Clean` backslashes them on Windows while `folder.path` is stored with forward slashes (`model.NewFolder` uses `path.Split`), so both descendant predicates matched nothing there. The ID arm still matched because `model.FolderID` applies the same `filepath.Clean` on both sides, exactly as you described. Switched to `path.Clean` (same normalization `HasAudioOutsideFolders` already uses) and added a nested-path test; verified it fails with backslashed predicates. Note `getFolderUpdateInfoBatch` in the same file has the same latent issue for nested selective-scan targets (its input is also slash-form via `libraryRelativePath`) - it predates this PR and matches the existing `#TBD-path-sep-scanner` skip markers, so I'm leaving it for the path-separator cleanup rather than widening this PR.

File: scanner/phase_1_folders.go [hunk 2]
```diff
@@ -518,6 +527,7 @@ func (p *phaseFolders) finalize(err error) error {
 		}
 		return nil
 	}, "scanner: finalize phaseFolders")
+	p.enqueueArtworkForImageChanges()
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/phase_1_folders.go]: The mechanism is real but it predates this PR: the finalize path (mark missing + TouchByMissingFolder, no artwork enqueue) is unchanged from master except for the added `enqueueArtworkForImageChanges()` call, and phase 3's refreshAlbum is only an `Album.Put`. It's also not indefinite: the album's state row points at the deleted file, so the first request that touches the original (any full-size `getCoverArt`, or any size not already in the resize cache) fails `os.Open`, hits `dangling()` and re-enqueues. The warm-resize-cache window is the same known limitation the serve-time self-heal always had. Collecting album/artist IDs from missing folders before `MarkMissing` is risky at that point: `finalize` sees every missing folder, so a temporarily unmounted library would bulk re-enqueue the whole collection with zero local candidates and go straight to the external agents (we've had that incident class before), and genuinely deleted albums would get pointless network lookups since their rows survive GC. If we want scan-time coverage for this, it needs to be scoped to albums that still have at least one non-missing folder, which is a separate change - tracking it as a follow-up rather than folding it in here.

**Comment by deluan** [comment]
@codex review

File: scanner/folder_entry.go [hunk 3]
```diff
@@ -69,6 +75,21 @@ func (f *folderEntry) isOutdated() bool {
...
+// state, and whether an artist image is involved (present in the old or the new list).
+func (f *folderEntry) imagesChanged() (changed, artistImage bool) {
+	newNames := slices.Sorted(maps.Keys(f.imageFiles))
+	prevNames := slices.Sorted(slices.Values(f.prevImageFiles))
+	// Both empty also skips the timestamp check, which is noise for image-less folders.
+	if len(prevNames) == 0 && len(newNames) == 0 {
+		return false, false
+	}
+	if slices.Equal(prevNames, newNames) && f.prevImagesUpdatedAt.Equal(f.imagesUpdatedAt) {
+		return false, false
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/folder_entry.go:88]: Confirmed the mechanism, with three corrections after digging in. It's narrower and wider than described: multiple images aren't required (the directory's own mtime is a floor in `imagesUpdatedAt`, so a single-image folder misses too), and GNU rsync on Linux does NOT trigger it (temp+rename bumps the dir mtime, which is detected) - the real triggers are `cp -p`, `rsync --inplace`, and macOS rsync writing in place with a restored older mtime. Also, a full scan doesn't repair it either (`EnqueueIfMissing` skips resolved items by design). But: this class is exactly as broken on master (which has no scan-side image enqueue at all), and the `ref_mtime` invariant self-heals it - an older mtime mismatches just like a newer one, so the first full-size or cache-cold read goes dangling and re-enqueues. Only warm resize-cache sizes serve stale, bounded by eviction. Both cheap fixes are traps: treating any outdated image folder as changed re-enqueues the whole library on every full scan (the config-fingerprint incident shape, and it breaks the repeat-full-scan guard tests), and dropping `folder.modTime` from `imagesUpdatedAt` changes the timestamp semantics so the first post-upgrade scan flags every image folder. The precise fix is a persisted per-image signature (`images_hash` column, empty value falling back to the current gate to avoid the upgrade wave) - that's a migration and belongs in its own PR. Tracking it as a follow-up.

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

File: scanner/phase_1_folders.go [hunk 4]
```diff
@@ -339,6 +341,19 @@ func (p *phaseFolders) persistChanges(entry *folderEntry) (*folderEntry, error)
 		albumRepo := tx.Album(p.ctx)
 		mfRepo := tx.MediaFile(p.ctx)
 
+		// A new folder's albums/artists are enqueued below; only pre-existing folders need the diff.
+		if !entry.isNew() {
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/phase_1_folders.go:347]: Confirmed and fixed in 3d7ac75dd, though I took the fix a level deeper than suggested. Reproduced it exactly as you described: full scan, then a quick scan that adds a new artist's album (the artist folder holds only subfolders, so `hasNoFiles()` made the producer skip it and it never got a row), then adding `artist.jpg` - the entry is `isNew()`, the diff is skipped, and with no tracks there is no inline enqueue either. Nothing queued.

Rather than recording new image-only folders (which reaches `enqueueArtworkForImageChanges`, and that uses plain `Enqueue`, so a first full scan would re-resolve ~1200 already-resolved items on the prod-shaped library I measured), I fixed the root cause in the producer: skip only `isEmpty()` new folders instead of `hasNoFiles()` ones. Genuinely empty directories are still skipped, ancestor folders get the row a full scan would already have given them, and the `!entry.isNew()` guard keeps its original rationale untouched. Bonus: this also fixes artist artwork settling absent for artists first imported by a quick scan, since the resolver's folder climb needs that same row - a separate gap I'd hit in live testing.

Added the quick-scan lifecycle test you asked for and verified it bites: it fails with `hasNoFiles()`, passes with `isEmpty()`.

File: persistence/folder_repository.go [hunk 5]
```diff
@@ -172,23 +179,68 @@ func (r folderRepository) getFolderUpdateInfoBatch(lib model.Library, targetPath
 
 // queryFolderUpdateInfo executes the query and returns the result map
 func (r folderRepository) queryFolderUpdateInfo(where And) (map[string]model.FolderUpdateInfo, error) {
-	sq := r.newSelect().Columns("id", "updated_at", "hash").Where(where)
+	sq := r.newSelect().Columns("id", "updated_at", "hash", "image_files", "images_updated_at").Where(where)
```
[the remark this answers is not in the document]
Comment (deluan) [reply persistence/folder_repository.go:172]: Fixed in 3d7ac75dd, but correcting the severity: the mechanism is real, the stated consequence isn't. When descendants are missing from `lastUpdates` they come back as a zero `FolderUpdateInfo`, so `isNew()` and `isOutdated()` are both true and they get fully re-processed - their tracks re-import and `persistChanges` force-enqueues their albums and artists inline (plain `Enqueue` on a selective scan, not `EnqueueIfMissing`). So both scenarios this PR targets are still covered on Windows via that fallback: the multi-disc cover through CD1/CD2 re-importing the same album, and the artist image through the album subfolders re-enqueueing the same artists. Nothing is missed or corrupted - the new enqueue is just silently inoperative there, falling back to master's behavior.

That said, you're right that it's inconsistent to add `folderSubtreeFilter` with a comment about exactly this hazard and leave a `filepath.Clean` fifty lines above it, so I switched the preload to `path.Clean` too (`os` and `path/filepath` are now unused in the file and dropped). Worth noting the four nested-path specs there are already `SkipOnWindows("path storage (#TBD-path-sep-persistence)")`, which is why Windows CI never caught either one.

**Comment by deluan** [comment]
@codex review

File: persistence/album_repository.go [hunk 6]
```diff
@@ -270,6 +270,30 @@ func (r *albumRepository) GetAllIDs(options ...model.QueryOptions) ([]string, er
...
+
+// GetSoleAlbumArtistIDsInSubtrees resolves the affected artists in one statement: matching albums
+// by their own folder_ids is both the resolver's notion of an album's folders and ~5x faster than
+// walking media_file rows for the subtree.
+func (r *albumRepository) GetSoleAlbumArtistIDsInSubtrees(lib model.Library, paths ...string) ([]string, error) {
+	if len(paths) == 0 {
+		return nil, nil
+	}
+	inSubtree := Exists("json_each(album.folder_ids) je join folder on folder.id = je.value",
+		folderSubtreeFilter(lib, paths))
```
[the remark this answers is not in the document]
Comment (deluan) [reply persistence/album_repository.go]: Confirmed and fixed in b6335a198, and the ceiling is lower than estimated: I measured it at **166 paths**, not ~200 (165 OK, 166 fails with exactly the error you quoted). The prod-shaped library I benchmark against has **158** folders matching the default `artist.*` - one artist short of tripping it. Adding `folder.*` to ArtistArtPriority would take that to 6,134.

This is a regression I introduced: the earlier `Folder.GetSubtreeIDs` chunked at 300 conditions with a comment about this exact limit, and collapsing it into a single statement dropped the chunking. Restored at 100 paths per chunk (deliberately well clear - the fixed depth from the EXISTS/AND wrapper counts too, and a 4th condition per path would silently drop the ceiling to ~124). Duplicate artist IDs across chunks are harmless since the queue upserts by PK.

You were also right that the album items were lost with it - `collectImageChangeQueueItems` returned the error after accumulating them. The artist failure is now logged and the collected album items are still returned, so any future limit surprise degrades to 'albums enqueued, artists missed' rather than nothing.

Regression test asserts 200 paths succeed; verified it fails with chunking disabled.

File: core/artwork/resolve.go [hunk 7]
```diff
@@ -525,6 +519,23 @@ func resolveFolderFile(ctx context.Context, lib libraryView, imgFiles []string,
...
+// ArtistArtPriority, so callers outside the resolver share the chain's token grammar.
+func IsArtistImageFile(name string) bool {
+	name = strings.ToLower(name)
+	for pattern := range strings.SplitSeq(strings.ToLower(conf.Server.ArtistArtPriority), ",") {
+		pattern = strings.TrimSpace(pattern)
+		if pattern == "" || pattern == externalCandidate || pattern == "image-folder" {
+			continue
+		}
+		pattern = strings.TrimPrefix(pattern, "album/")
+		if ok, _ := path.Match(pattern, name); ok {
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/artwork/resolve.go]: Confirmed and fixed in b6335a198. Verified the resolver really does support these: `findImageInFolder` builds `path.Join(escapeGlobLiteral(relFolder), pattern)` and only escapes the folder part, so the pattern's `/` stays a separator and `fs.Glob` genuinely matches `Artist/images/artist.jpg`. Confirmed the rest of your chain too - the scanner passes bare basenames (`folder.image_files` stores names, not paths), and that folder has no tracks, so nothing else enqueues the artist.

Also a regression of mine: my first version used `path.Base(pattern)` and I changed it to `TrimPrefix(pattern, "album/")` in a cleanup pass to mirror the token grammar more exactly. Reverted to `path.Base`, which is a net simplification - `path.Base("album/artist.*")` is already `artist.*`, so behavior is byte-identical for every default config, and only multi-segment patterns diverge.

Went with the conservative basename match rather than the precise one because an exact classifier isn't computable in the scanner: the artist folder is derived at resolve time from the common parent of the artist's albums, and `fromArtistFolder` climbs up to 3 levels, so the same file has a different relative prefix per artist. The over-trigger costs one redundant re-resolve that produces the same content hash; the false negative costs permanently stale artwork with no repair path. Fail-open is the right direction for a change detector, and I noted that reasoning in the comment. Regression test added with `images/artist.*`; verified it fails with the old form.

**Comment by deluan** [comment]
@codex review

File: scanner/phase_1_image_changes.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	}
+}
+
+func (p *phaseFolders) collectImageChangeQueueItems(lib model.Library, folders []imageChangedFolder) ([]model.ArtworkQueueItem, error) {
+	folderIDs := make([]string, len(folders))
+	var artistFolderPaths []string
+	for i, f := range folders {
+		folderIDs[i] = f.id
+		if f.artistImage {
+			artistFolderPaths = append(artistFolderPaths, f.path)
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/phase_1_image_changes.go]: Your reading of the code is right - I reproduced it, and it's worse than described: replacing `Artist/images/artist.jpg` with `images/artist.*` configured enqueues *nothing*, not just no artist (the album branch finds nothing either, since that folder has no tracks and isn't a parent of the album folder).

But I'm not fixing it here, because nested artist patterns are already non-functional end to end, independently of this PR. `folders_artist.go` does `_, name := path.Split(p); return f, filepath.Join(absFolder, name)`, dropping the pattern's directory component - so the resolution reports `sourcePath=/music/Artist/artist.jpg` for a file at `/music/Artist/images/artist.jpg`. Measured:

```
real file : .../Artist/images/artist.jpg
sourcePath: .../Artist/artist.jpg
refMtime  : 0
```

`source="folder"` is file-backed, so nothing is written to the image store and the bogus path is persisted. Every later request hits `os.Open` → ENOENT → `dangling()` → placeholder and a re-enqueue that fails identically. So a nested-pattern artist image never survives a round trip, on master too. Making the scanner enqueue for it would only let the worker re-persist an unservable state row.

Worth noting the default config never reaches this: `album/artist.*` is a token special-cased to the album's own image list, not a directory-bearing glob, so it never goes through `fromArtistFolder`.

I also checked your "conservatively query the parent subtree" suggestion and it's actively unsafe: a depth-1 artist folder's parent is `.`, which `folderSubtreeFilter` treats as the whole library - so every artist-image change would fan out to every artist. Verified it breaks the per-artist isolation test. If the resolver bug is fixed later, the correct scanner change is to walk up the pattern's directory depth instead.

Tracking both as follow-ups: the `folders_artist.go` source-path bug first, then this subtree root. Keeping the basename classifier as-is - it's deliberately a superset.

**Comment by deluan** [comment]
@codex review

File: scanner/phase_1_folders.go [hunk 9]
```diff
@@ -339,6 +343,19 @@ func (p *phaseFolders) persistChanges(entry *folderEntry) (*folderEntry, error)
...
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/phase_1_folders.go:347]: This one is already handled, by the same commit you're referring to. The producer change in 3d7ac75dd isn't only forward-looking: it also self-heals the pre-upgrade state, so no migration is needed.

A quick scan walks the whole tree (`walkDirTree` recurses unconditionally; outdatedness is decided after the entry is built). A row-less ancestor gets a zero `FolderUpdateInfo`, so `prevHash == ""` != `hash()` and it is outdated. It is no longer skipped, because the guard is now `isEmpty()` (`hasNoFiles() && numSubFolders == 0`) and an artist folder holding album subfolders has subfolders. So `persistChanges` creates the row on the first post-upgrade scan, before any `artist.jpg` can be added.

Verified empirically: built `Kraftwerk/Autobahn/track.mp3`, deleted the `Kraftwerk` folder row via SQL to reproduce exactly the pre-upgrade state, then ran a quick scan - the row came back. Adding `artist.jpg` and scanning again enqueued the `ar` item. Control: flipping the guard back to `hasNoFiles()` makes it fail at the self-heal assertion, which is what would have happened on master.

On the startup-scan point: `scanonstartup` defaults to true and passes no targets, so upgrading (which requires a restart) runs a full-tree quick scan before the user can touch anything.

There is a narrower residual you didn't describe, which I'm leaving as a follow-up: with `ScanOnStartup=false` and no schedule, the first post-upgrade scan can be the selective watcher scan triggered by the `artist.jpg` write itself. Then the row is created and the image seen in the same scan, so `isNew()` is true and the record is suppressed. That needs non-default config and is not a regression - master enqueued nothing for image changes at all.

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

**Comment by deluan** [comment]
@codex review

File: scanner/phase_1_folders.go [hunk 10]
```diff
@@ -339,6 +343,15 @@ func (p *phaseFolders) persistChanges(entry *folderEntry) (*folderEntry, error)
...
```
[the remark this answers is not in the document]
Comment (deluan) [reply scanner/phase_1_folders.go:347]: Traced every default-config path and this produces no missed enqueue, so I'm closing it like the earlier nested-glob one.

The mechanism you describe is real but self-limiting. A brand-new directory means every descendant path is new too, so `processFolder` finds no existing tracks, imports all of them, and `createAlbums/ArtistsFromMediaFiles` runs - which means the inline loops in `persistChanges` already enqueue every album and artist in that subtree. The set the suppressed record could add is always a subset of what was just enqueued inline. The one case where it isn't a subset is an image-only folder with no albums beneath it, and there the set is empty.

Checked concretely, all passing as throwaway specs: new `Artist/` with `artist.jpg` plus album subfolders (artist and album both enqueued, inline from the album folder); new `Artist/Album/` with `cover.jpg` and tracks (album enqueued); tracks only in `CD1/` under an image-only album root (both enqueued); existing folder gaining its first `artist.jpg` (enqueued, covered by the shipped test); an album moved into a new folder carrying `artist.jpg` (both enqueued). The one case that enqueues nothing is a new image-only folder like `The Beatles/Scans/front.jpg` - and nothing should be enqueued, because the resolver can never read it: `fromArtistFolder` globs the artist folder and climbs up at most 3 ancestors, never into a sibling, and the album readers use the album's own folders plus their common parent.

I also tested your suggested fix directly by replacing the guard with `if true`: zero behavioural difference across every default-config scenario and the whole image-change suite. It adds no queue item, and on a first scan - where every folder is new - it would record every image-bearing folder and run library-sized subtree queries producing duplicates of what the inline path already queued. That's the redundant-wave shape I rejected in an earlier round, for no gain.

As you note, the nested `images/artist.*` premise also depends on the separate mapping fix - and that configuration can't serve artwork at all today, since `folders_artist.go` drops the pattern's directory from the persisted source path, so those images return a placeholder on every request forever. Both are tracked as follow-ups.
